cmd/gke-certificates-controller/app/gke_approver.go

```diff
@@ -170,34 +189,151 @@ func hasExactUsages(csr *capi.CertificateSigningRequest, usages []capi.KeyUsage)
+func isNodeCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
 	if !reflect.DeepEqual([]string{"system:nodes"}, x509cr.Subject.Organization) {
 		return false
 	}
 	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
 		return false
 	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+		return false
+	}
+	return true
+}
+
+func isNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
 	if !hasExactUsages(csr, kubeletClientUsages) {
 		return false
 	}
-	if !strings.HasPrefix(x509cr.Subject.CommonName, "system:node:") {
+	return true
+}
+
+func isSelfNodeClientCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeClientCert(opts, csr, x509cr) {
+		return false
+	}
+	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
 
-func isSelfNodeClientCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
-	if !isNodeClientCert(csr, x509cr) {
+func isNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletServerUsages) {
 		return false
 	}
 	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
+
+func validateNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+
+	if err := validateNodeServerCertInner(client, opts, csr, x509cr); err != nil {
+		glog.Errorf("validating CSR %q: %v", csr.Name, err)
+		return false
+	}
+	return true
+}
+
+// Only check that DNS name and IP in SAN match an existing VM in the project.
+// Username was already checked against CN, so this CSR is coming from
+// authenticated kubelet.
+func validateNodeServerCertInner(client *http.Client, opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	if len(x509cr.DNSNames) == 0 {
+		return errors.New("no SAN DNS names")
+	}
+	if len(x509cr.IPAddresses) == 0 {
+		return errors.New("no SAN IP addresses")
+	}
+	cs, err := compute.New(client)
+	if err != nil {
+		return fmt.Errorf("creating GCE API client: %v", err)
+	}
+
+	var errorMsgs []string
+	for _, n := range x509cr.DNSNames {
+		inst, err := compute.NewInstancesService(cs).Get(opts.projectID, opts.zone, n).Do()
+		if err != nil {
+			errorMsgs = append(errorMsgs, fmt.Sprintf("DNS name %q doesn't match any VM in cluster project/zone: %v", n, err))
+			continue
+		}
+		for _, iface := range inst.NetworkInterfaces {
+			for _, ip := range x509cr.IPAddresses {
+				if ip.String() == iface.NetworkIP {
+					return nil
```

Comment:
why return? don't we meed to validate all ips in the CSR, not just the first?

Reply:
You're right.
For some reason I thought "if any dns/ip combo passes it's sufficient". Brain fart.